### PR TITLE
Improve CI workflow

### PR DESCRIPTION
**Pull request by baptiste0928**
Hello! I noticed that your current CI workflow takes a lot of time to run. This PR improves it:

- Each step has been separated in different jobs (with matrix jobs for tests). This allows to run task concurrency, and makes the workflow easier to understand.
- Dependencies are cached with [Swatinem/rust-cache](https://github.com/Swatinem/rust-cache) and [baptiste0928/cargo-install](https://github.com/baptiste0928/cargo-install) to speed up workflows.
- Clippy is also run on tests (I fixed few lints)

image

With these improvements, the workflow takes only 12 minutes to complete, vs. half an hour with the current one.

This PR contains a lot of useless commits, I suggest you to squash-merge the PR instead of a regular merge.

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -10,51 +10,163 @@ env:
...
+      - name: Checkout sources
+        uses: actions/checkout@v2
+
+      - name: Run cargo fmt
+        run: cargo fmt --all -- --check
+
+  # Tests
+  test:
+    name: Test
+    needs: [msrv]
```
Comment (surban) [inline .github/workflows/rust.yml:84]: I don't understand why this depends on `msrv`. Are the tests run with current stable Rust or with MSRV Rust?

**Review by surban — COMMENTED** [review]
Thanks! This looks great. I was not aware that you can do such things with GitHub actions.

Only I don't really understand the dependencies between the jobs.
Shouldn't the MSRV task be independent of all other tasks?

**Comment by baptiste0928** [comment]
I did this to make sure there are no compilation errors before running the other jobs. It allows to know faster that something is not compiling, and avoids to launch useless jobs.

It can be changed though, there is nothing that forces this dependency.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (baptiste0928) [reply .github/workflows/rust.yml:84]: All jobs are run in a fresh environment, so all jobs except `msrv` uses the current stable Rust (included in the base ubuntu-latest image). Setting a different default version in `msrv` doesn't affect other jobs, even if they depend on `msrv`.

**Pull request merged**